src/EditorFeatures/CSharp/CompleteStatement/CompleteStatementCommandHandler.cs

```diff
@@ -0,0 +1,426 @@
+                // ArgumentList: covers method invocations like object.Method(arg)
+                // ArrayRankSpecifier: covers new Type[dim]
+                // ElementAccessExpression: covers indexer invocations like array[index]
+                // ParenthesizedExpression: covers (3*(x+y))
+                if (currentNode.IsKind(SyntaxKind.ArgumentList, SyntaxKind.ArrayRankSpecifier, SyntaxKind.ElementAccessExpression, SyntaxKind.ParenthesizedExpression))
+                {
+                    nodeFound = true;
+                }
+
+                // No special action is performed at this time if `;` is typed inside a string, including
+                // interpolated strings.  
+                if (IsInAString(currentNode, isCaretAtEndOfLine))
+                {
+                    return false;
+                }
+
+                // We reached the root without finding a statement
+                if (currentNode.Parent == null)
+                {
+                    return false;
+                }
+
+                currentNode = currentNode.Parent;
+            }
+
+            // if we never found a statement, or a node of interest, or the statement kind is not a candidate for completion, return
+            if (currentNode == null || !nodeFound || !StatementIsACandidate(currentNode, caret))
+            {
+                return false;
+            }
+
+            Debug.Assert(currentNode != null);
+            return true;
+        }
+
+        private static bool IsStatementOrFieldDeclaration(SyntaxNode currentNode, ISyntaxFactsService syntaxFacts)
+            => syntaxFacts.IsStatement(currentNode) || currentNode.IsKind(SyntaxKind.FieldDeclaration);
+
+        private static bool IsInAString(SyntaxNode currentNode, bool isCaretAtEndOfLine)
+                     // If caret is at the end of the line, it is outside the string	
+                     => (currentNode.IsKind(SyntaxKind.InterpolatedStringExpression, SyntaxKind.StringLiteralExpression)
+                         && !isCaretAtEndOfLine);
+
+        private static bool StatementIsACandidate(SyntaxNode currentNode, int caretPosition)
+        {
+            // if the statement kind ends in a semicolon, return true
+            switch (currentNode.Kind())
+            {
+                case SyntaxKind.DoStatement:
+                case SyntaxKind.ExpressionStatement:
+                case SyntaxKind.GotoCaseStatement:
+                case SyntaxKind.LocalDeclarationStatement:
+                case SyntaxKind.ReturnStatement:
+                case SyntaxKind.ThrowStatement:
+                case SyntaxKind.FieldDeclaration:
+                    return true;
+                case SyntaxKind.ForStatement:
+                    var forStatementSyntax = (ForStatementSyntax)currentNode;
+                    if (CaretIsInForStatementCondition(caretPosition, forStatementSyntax) ||
+                        CaretIsInForStatementDeclaration(caretPosition,forStatementSyntax))
```

Comment:
1. needs to handle 'Initializers' case for a For-Statement.
2. having an 'If-statemnet' that returns 'true' in the when-true part, and 'false' in the when-false part can just be simplified to:

```return CaretIsIn...(...) || CaretIsIn...(...);```

Reply:
Fixed